### ISSUE TITLE
Reference-listing path for draft-arkko-farrell-arch-model-t steers to XML instead of document in output files

### ISSUE THREAD

**Issue by lbartholomew-rpc** [ietf-tools/bibxml-service#378]
### Describe the issue

Path in the XML file is typical:
https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.arkko-farrell-arch-model-t.xml

but for draft-arkko-farrell-arch-model-t, the link is unusual -- includes "api/v1/":
Output for
https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.arkko-farrell-arch-model-t.xml
on https://www.rfc-editor.org/authors/rfc9464.html is
https://datatracker.ietf.org/api/v1/doc/document/draft-arkko-farrell-arch-model-t/

as compared to output for
https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.ietf-tls-esni.xml on
https://www.rfc-editor.org/authors/rfc9462.html
which is as it should be:
https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17

### Code of Conduct

- [X] I agree to follow the [IETF's Code of Conduct](https://github.com/ietf-tools/.github/blob/main/CODE_OF_CONDUCT.md)

**Comment by lbartholomew-rpc** [comment]
Forgot to add that the file that contains the problem draft link is RFC-to-be 9464, which is otherwise ready to be published as an RFC, so publication will have to be delayed until (1) this issue is fixed or (2) we ask the authors if they want to remove draft-arkko-farrell-arch-model-t from the references list and they decide to do so.

**Comment by rjsparks** [comment]
The source of truth at https://datatracker.ietf.org/doc/bibxml3/reference.I-D.arkko-farrell-arch-model-t.xml points to the correct place.

**Comment by rjsparks** [comment]
@lbartholomew-rpc - consider changing the source document to point to https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.draft-arkko-farrell-arch-model-t-04.xml instead

**Comment by lbartholomew-rpc** [comment]
Hi, Robert.
Re. "The source of truth at https://datatracker.ietf.org/doc/bibxml3/reference.I-D.arkko-farrell-arch-model-t.xml points to the correct place." -- this didn't work for me.

Re. pointing to  https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.draft-arkko-farrell-arch-model-t-04.xml instead -- this worked.  Is there something special that we need to know re. whether or not to include "draft-" and/or the version number?  Thanks.

**Comment by rjsparks** [comment]
leaving out both draft- and the version number gives you a pointer to a pronoun that will resolve to whatever the current version is at the time you reference it. Providing draft- and the version number will always resolve to that specific version.

**Comment by lbartholomew-rpc** [comment]
Good to know; thanks!

**Comment by kesara** [comment]
@stefanomunarini Do you have any insight into the `target` in https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.arkko-farrell-arch-model-t.xml has the API call URL?

**Comment by stefanomunarini** [comment]
@kesara I guess this question should be asked to the Datatracker team, instead.

The bibxml-service simply sends a request to the Datatracker APIs and extracts this piece of data from the response without any sort of manipulation whatsoever (see [here](https://github.com/ietf-tools/bibxml-service/blob/31dc5f297ad8be097e6f861fc248c386b74cd914/datatracker/internet_drafts.py#L90)).

**Comment by kesara** [comment]
@stefanomunarini there must be somewhere else in the BibXML service which manipulates this URL.
For example, https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.cuiling-dnsop-sm2-alg.xml has recent drafts. This has the correct `target` URL.

https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.arkko-farrell-arch-model-t.xml has the following in the header:
```
x-resolution-outcomes: ;InternetDraftsAdapter: using query (@.type == "Internet-Draft") && (@.id like_regex "draft\-arkko\-farrell\-arch\-model\-t\-[[:digit:]]{2}") -> returning Datatracker version 04,;
```

While https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.cuiling-dnsop-sm2-alg.xml has
```
x-resolution-outcomes: ;InternetDraftsAdapter: using query (@.type == "Internet-Draft") && (@.id like_regex "draft\-cuiling\-dnsop\-sm2\-alg\-[[:digit:]]{2}"),;
```

For some reason, BibXML thinks there's a newer version in datatracker. See https://github.com/ietf-tools/bibxml-service/blob/main/bibxml/xml2rfc_adapters.py#L295

BTW if BibXML thinks there's a newer version, BibXML shouldn't return the API URL rather use the data from API call to get the latest URL.

**Comment by rjsparks** [comment]
@stefanomunarini Read further up issue. The datatracker is _NOT_ returning the api as the target as the links already above show.

**Comment by stefanomunarini** [comment]
> For some reason, BibXML thinks there's a newer version in datatracker. See https://github.com/ietf-tools/bibxml-service/blob/main/bibxml/xml2rfc_adapters.py#L295
> 
> BTW if BibXML thinks there's a newer version, BibXML shouldn't return the API URL rather use the data from API call to get the latest URL.

You're right, In fact the bibxml-service is returning this warning when retrieving this draft:

`Returning Datatracker result for xml2rfc bibxml3 path. If unversioned I-D was requested, then Datatracker may have a newer I-D version than indexed sources. Alternatively, indexed version could not be used for some reason. Requested version None, indexed sources have version 02, returning Datatracker’s version 04.`

bibxml-service does not manipulate the value itself, however it uses a callback to retrieve the most recent version of an ID document, using the Datatracker APIs.

In this particular case, Datatracker has version 4 of this Internet Draft, while the service is only indexing version 2 of the same draft. 

I am investigating why this is happening, and also why the Datatracker API URL is used in these cases.

**Comment by strogonoff** [comment]
It looks like we may be matching `draft-arkko-farrell-arch-model-t-<number>-additions[-<number>]` by mistake. The regular expression by this link from @stefanomunarini may be missing end of string marker:

https://github.com/ietf-tools/bibxml-service/blob/31dc5f297ad8be097e6f861fc248c386b74cd914/bibxml/xml2rfc_adapters.py#L183-L187

If it was `[[:digit:]]{2}$` then it would only match `draft-arkko-farrell-arch-model-t-<two digits>` .

**Comment by strogonoff** [comment]
We should also clarify in the code that the two digits are intended to represent possible versions.

**Comment by TonyLHansen** [comment]
should the {2} be {2-3} (or whatever the syntax is) ? I can't find any examples right now of drafts that went over 99, but I definitely found examples of ones that went to 99.

**Comment by rjsparks** [comment]
No - we've never had > 99 in the repo/archive to my knowledge. And the datatracker only knows things up to 99.
I'm pretty sure the IESG forced a rename before going to 100 in the most recent case you may remember.

That said, trying to find versions at the end of drafts by regular expression alone is almost impossible. There is some backtracking you have to do (we've done it with iteration in python) because things like draft-foo-bar-802-11-00 exist.

**Comment by stefanomunarini** [comment]
> draft-foo-bar-802-11-00 

https://github.com/ietf-tools/bibxml-service/blob/31dc5f297ad8be097e6f861fc248c386b74cd914/bibxml/xml2rfc_adapters.py#L183-L187

In this case, `draft-foo-bar-802-11` is the name of the document (variable `unversioned` in the snippet). With the regex we are able to match the version(s) at the end of the string. 

Or am I not getting something else here? @rjsparks

**Comment by rjsparks** [comment]
How do you deal with the string `draft-ietf-tsvwg-ieee-802-11`? Do you try to treat that as version 11 of `draft-ietf-tsvwg-ieee-802`? Or is something keeping the string from getting to that regex?

More concretely, see https://bib.ietf.org/public/rfc/bibxml-ids/reference.I-D.ietf-tsvwg-ieee-802.xml and note that the returned anchor is incorrect - there is no such thing as `I-D.ietf-tsvwg-ieee-802`.

**Comment by alicerusso** [comment]
This issue seems to currently affect
https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.ietf-regext-rdap-geofeed.xml
which contains an incorrect target:
`target="https://datatracker.ietf.org/api/v1/doc/document/draft-ietf-regext-rdap-geofeed/"`

**Comment by rjsparks** [comment]
That draft is not falling victim to the version matching problem, but is still spitting out a reference to the v1 api. @stefanomunarini what did you discover about where that's coming from.
It's not coming from the datatracker's feed to the bibxml service - see https://datatracker.ietf.org/doc/bibxml3/reference.I-D.ietf-regext-rdap-geofeed.xml
